File: src/nl/weeaboo/kid/ever17/StringCommandReader.java
package nl.weeaboo.kid.ever17;

import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;

import nl.weeaboo.kid.KIDUtil;
import nl.weeaboo.vnds.Log;

public class StringCommandReader {

   private ByteBuffer input;
   private final StringBuilder output;

   public StringCommandReader() {
      output = new StringBuilder();
   }

   //Functions
   public String parse(ByteBuffer in) throws UnsupportedEncodingException {
      setInput(in);
      parse();

      return output.toString();
   }

   public void parse() throws UnsupportedEncodingException {
      output.delete(0, output.length());

      int c;
      OUTER:
      while ((c = peek()) != 0) {
         StrOpcode op = StrOpcode.get(c);
         if (op == null) {
            output.append(readChar());
         } else {
            c = read();
            switch (op) {
               case waitForClick:
                  output.append(String.format("{%s}", op));
                  break;
               case clearText:
                  output.append(String.format("{%s}", op));
                  break;
               case delay:
                  output.append(String.format("{%s %s}", op, readExpr()));
                  break;
               case appendText:
                  output.append(String.format("{%s}", op, readExpr()));
                  break;
               case sound:
                  output.append(String.format("{%s %s}", op, readCString()));
                  break;
               case waitForSound:
                  output.append(String.format("{%s}", op));
                  break;
               case choice: {
                  int arg0 = read();
                  int arg1 = readShort();
                  output.append(String.format("{%s %02x %04x ", op, arg0, arg1));
                  while (peek() != 0) {
                     String option = parseChoiceOption();
                     output.append('|');
                     output.append(option);
                  }
                  read(); //Read away the trailing 0
                  output.append("}");
                  break OUTER;
               }
               //pout.printf("{%s}", op);
               case marker:
                  output.append(String.format("{%s}", op));
                  break;
               case nextPage: {
                  int arg0 = read();
                  output.append(String.format("{%s %02x}", op, arg0));
                  break;
               }
               case bigChar:
                  output.append(String.format("{%s}", op));
                  break;
               default:
                  break;
            }
         }
      }
   }

   protected String parseChoiceOption() throws UnsupportedEncodingException {
      int marker = read();
      if (marker != 0x0B) {
         Log.v("     [Unknown Text] Expected choice-option's marker to be: " + 0x0B + ", but got: " + marker);
      }

      int type = read();

      StringBuilder temp = new StringBuilder();
      if (type == 1) {
         var c = readChar();
         while (!c.equals("\n")) {
            temp.append(c);
            c = readChar();
         }
      } else if (type == 2) {
         int arg0 = read();
         int arg1 = read();
         int arg2 = read();
         int varname = read();
         int op = read();
         int arg3 = read();
         if (op != 0x14) {
            Log.v("     [Unknown Text] Unknown operator in choice option cond: " + op);
         }
         temp.append(String.format("{%s (%02x %02x %02x) %02x %02x (%02x)}",
                 "cond", arg0, arg1, arg2, varname, op, arg3));
         var c = readChar();
         while (!c.equals("\n")) {
            temp.append(c);
            c = readChar();
         }
      }
      return temp.toString();
   }

   //Getters
   protected int peek() {
      if (!input.hasRemaining()) {
         return 0;
      }
      return input.get(input.position()) & 0xFF;
   }

   protected int read() {
      return input.get() & 0xFF;
   }

   protected int readShort() {
      return input.getShort() & 0xFFFF;
   }

   protected int readInt() {
      return input.getInt() & 0xFFFFFFFF;
   }

   private final byte[] readCharTemp = new byte[4];

   protected String readChar() throws UnsupportedEncodingException {
      int c = read();
      if (c == 0x01) {
         return "\n";
      }

      int numBytes = 1;
      // Ever17 JP Edition and Ever1 US Edition are all using SJIS encoding
//      if (inputEncoding.equals("SJIS")) {
      if ((c >= 0x80 && c <= 0xa0) || (c >= 0xe0 && c <= 0xef)) {
         numBytes = 2;
      }

      readCharTemp[0] = (byte) c;
      for (int n = 1; n < numBytes; n++) {
         readCharTemp[n] = (byte) read();
      }

      var rs = new String(readCharTemp, 0, numBytes, "MS932");
      switch (rs) {
         case "⑩":
            rs = "ä";
            break;
         case "⑪":
            rs = "ö";
            break;
         case "⑫":
            rs = "ü";
            break;
         case "⑬":
            rs = "—"; // long dash
            break;
      }

      return rs;
   }

   private String readCString() {
      return KIDUtil.readCString(input);
   }

   private String readExpr() {
      return Ever17Util.readExpr(input);
   }

   //Setters
   public void setInput(ByteBuffer in) {
      input = in;
   }

}

File: tools/converter/StringCommandReader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace converter {
   class StringCommandReader {
      private BinaryReader input;
      private readonly StringBuilder output;
      public StringCommandReader() {
         output = new StringBuilder();
      }
      public void SetInput(BinaryReader @in) {
         input = @in;
      }
      public string Parse(BinaryReader inp) {
         SetInput(inp);
         Parse();
         return output.ToString();
      }
      public void Parse() {
         output.Remove(0, output.Length);
         int c;
         while ((c = Peek()) != 0) {
            StrOpcode op = StrOpcode.Get(c);
            if (op == null) {
               output.Append(ReadChar());
            }
            else {
               c = Read();
               if (op == StrOpcode.waitForClick) {
                  output.Append($"{{{op}}}");
               }
               else if (op == StrOpcode.clearText) {
                  output.Append($"{{{op}}}");
               }
               else if (op == StrOpcode.delay) {
                  output.Append($"{{{op} {ReadExpr()}}}");
               }
               else if (op == StrOpcode.appendText) {
                  output.Append(string.Format("{{{0}}}", op, ReadExpr()));
               }
               else if (op == StrOpcode.sound) {
                  output.Append($"{{{op} {ReadCString()}}}");
               }
               else if (op == StrOpcode.waitForSound) {
                  output.Append($"{{{op}}}");
               }
               else if (op == StrOpcode.choice) {
                  int arg0 = Read();
                  int arg1 = ReadShort();
                  output.Append(string.Format("{{{0} {1:x2} {2:x4} ", op, arg0, arg1));
                  while (Peek() != 0) {
                     string option = ParseChoiceOption();
                     output.Append('|');
                     output.Append(option);
                  }
                  Read(); //Read away the trailing 0
                  output.Append("}");
                  goto OUTER;
               }
               else if (op == StrOpcode.marker) {
                  output.Append($"{{{op}}}");
               }
               else if (op == StrOpcode.nextPage) {
                  int arg0 = Read();
                  output.Append(string.Format("{{{0} {1:x2}}}", op, arg0));
               }
               else if (op == StrOpcode.bigChar) {
                  output.Append($"{{{op}}}");
               }
            }
         }
OUTER:;
      }
      protected string ParseChoiceOption() {
         int marker = Read();
         if (marker != 0x0B) {
            Log.Write("     [Unknown Text] Expected choice-option's marker to be: " + 0x0B + ", but got: " + marker);
         }

         int type = Read();

         StringBuilder temp = new StringBuilder();
         if (type == 1) {
            var c = ReadChar();
            while (c != "\n") {
               temp.Append(c);
               c = ReadChar();
            }
         }
         else if (type == 2) {
            int arg0 = Read();
            int arg1 = Read();
            int arg2 = Read();
            int varname = Read();
            int op = Read();
            int arg3 = Read();
            if (op != 0x14) {
               Log.Write("     [Unknown Text] Unknown operator in choice option cond: " + op);
            }
            temp.Append(string.Format("{{{0} ({1:x2} {2:x2} {3:x2}) {4:x2} {5:x2} ({6:x2})}}",
                    "cond", arg0, arg1, arg2, varname, op, arg3));
            var c = ReadChar();
            while (c != "\n") {
               temp.Append(c);
               c = ReadChar();
            }
         }
         return temp.ToString();
      }
      protected int Peek() {
         if (!input.BaseStream.HasRemaining()) {
            return 0;
         }
         int rs = input.ReadByte() & 0xFF;
         input.BaseStream.Position -= 1;
         return rs;
      }
      protected int Read() {
         return input.ReadByte() & 0xFF;
      }
      protected int ReadShort() {
         return input.ReadInt16() & 0xFFFF;
      }
      protected int ReadInt() {
         return input.ReadInt32();
      }
      private readonly byte[] readCharTemp = new byte[4];
      protected string ReadChar() {
         int c = Read();
         if (c == 0x01) {
            return "\n";
         }

         int numBytes = 1;
         // Ever17 JP Edition and Ever17 US Edition are all using SJIS encoding

         if ((c >= 0x80 && c <= 0xa0) || (c >= 0xe0 && c <= 0xef)) {
            numBytes = 2;
         }

         readCharTemp[0] = (byte)c;
         for (int n = 1; n < numBytes; n++) {
            readCharTemp[n] = (byte)Read();
         }
         
         var rs = Encoding.GetEncoding("shift_jis").GetString(readCharTemp, 0, numBytes);
         switch (rs) {
            case "①": // Circled Number One
               rs = "{💧💧}";
               break;
            case "②": // Circled Number Two
               rs = "❤️";
               break;
            case "③": // Circled Number Three
               rs = "💢";
               break;
            case "④": // Circled Number Four
               rs = "💦";
               break;
            case "⑤": // Circled Number Five
               rs = "★";
               break;
            case "⑩": // Circled Number Ten
               rs = "ä";
               break;
            case "⑪": // Circled Number Eleven
               rs = "ö";
               break;
            case "⑫": // Circled Number Twelve
               rs = "ü";
               break;
            case "⑬": // Circled Number Thirteen
               rs = "—"; // long dash
               break;
         }

         return rs;
      }
      private string ReadCString() {
         return KIDUtil.ReadCString(input);
      }

      private string ReadExpr() {
         return Ever17Util.ReadExpr(input);
      }

   }
}
